ENT-5768 startFlowWithClientId permissions

lankydan opened the pull request:
Do not let a user reattach to a flow started by another user.

Reattaching to a flow using `startFlowWithClientId` for a flow not
started by the current user throws a `PermissionException`

Reattaching to a flow using `reattachFlowWithClientId` for a flow not
started by the current user returns null.

`finishedFlowsWithClientIds` does not return flows started by other
users.

Normal rpc permissions around `startFlowWithClientId` and
`startFlowDynamicWithClientId` has also been added.

To allow admins to remove client ids as well as be able to see all the
client ids on the node, admin versions have been added that bypass the
user restrictions. These can be permitted via rpc to only provide
their usage to admins.

lankydan committed:
ENT-5768 startFlowWithClientId permissions

Do not let a user reattach to a flow started by another user.

Reattaching to a flow using `startFlowWithClientId` for a flow not
started by the current user throws a `PermissionException`

Reattaching to a flow using `reattachFlowWithClientId` for a flow not
started by the current user returns null.

`finishedFlowsWithClientIds` does not return flows started by other
users.

Normal rpc permissions around `startFlowWithClientId` and
`startFlowDynamicWithClientId` has also been added.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -320,6 +327,13 @@ internal class SingleThreadedStateMachineManager(
+                    // return an exception as they are not permitted to see the result of the flow
+                    if (!it.isPermitted(context.principal())) {
+                        return@startFlow openFuture<FlowStateMachineHandle<A>>().apply {
+                            setException(PermissionException("User not authorized to start flow with client id [$clientId]"))
```
I think we could put a bit of a hint in this message about it being a flow that is already running.

rick-r3 reviewed (commented):
I think we need to consider how some administrator would clean up unclaimed flow results.  Previously, without permissions an admin user could come and list them and remove them just fine.  Now they cannot.

Should we add two admin versions of list and remove that could be permissioned differently at the method level, that are able to list and remove any/all ?

lankydan committed:
ENT-5768 Admin version to skip user permissions

To allow admins to remove client ids as well as be able to see all the
client ids on the node, admin versions have been added that bypass the
user restrictions. These can be permitted via rpc to only provide
their usage to admins.

lankydan committed:
ENT-5768 Test running, completed and node restart

lankydan committed:
ENT-5768 Error message + detekt

lankydan committed:
ENT-5768 Remove reference to database object

lankydan committed:
ENT-5768 Change after rebase

rick-r3 commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -322,13 +323,33 @@ interface CordaRPCOps : RPCOps {
     /**
-     * Returns all finished flows that were started with a client id.
+     * Removes a flow's [clientId] to result/ exception mapping. If the mapping is of a running flow, then the mapping will not get removed.
+     * This version can be called for all client ids, ignoring which user originally started a flow using the input [clientId].
```
`flow using input` -> `flow with`

rick-r3 commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -322,13 +323,33 @@ interface CordaRPCOps : RPCOps {
     fun finishedFlowsWithClientIds(): Map<String, Boolean>
 
+    /**
+     * Returns all finished flows that were started with a client id. This version does not filter the returned values by the user that
```
`Returns all finished flows that were started with a client id by all RPC users for which the client ID mapping has not been removed.`

rick-r3 commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -322,13 +323,33 @@ interface CordaRPCOps : RPCOps {
+    fun removeClientIdAsAdmin(clientId: String): Boolean
+
+    /**
+     * Returns all finished flows that were started with a client id. This version only returns the client ids for flows started by the same
```
`Returns all finished flows that were started with a client ID for which the client ID mapping has not been removed. This version only returns the client ids for flows started by the same user currently calling [finishedFlowsWithClientIds].`

rick-r3 reviewed (dismissed):
Just some wording changes in the kdocs.

lankydan committed:
ENT-5768 Kdocs
